doc: contribution guidelines: Clarify and extend (#83117)
zephyrproject-rtos/zephyr: doc/contribute/guidelines.rst

Coding Style
============

.. note::
   Coding style is enforced on any new or modified code, but contributors are
   not expected to correct the style on existing code that they are not
   modifying.

.. _Linux kernel coding style:
   https://kernel.org/doc/html/latest/process/coding-style.html

.. _snake case:
   https://en.wikipedia.org/wiki/Snake_case

In general, follow the `Linux kernel coding style`_, with the following
exceptions and clarifications:

> * Use `snake case`_.

Reviewer: I this valid ? What about macro's ?

Author: This is yet another reason why I avoided this statement in the first place. In the Linux Kernel coding style, this is not mentioned at all, and is instead implicit. I will add a clarification about macros.

Reviewer 2: Can just change it to say snake case for variable and function names, therefore excluding defines and macros

Author: I clarified it a bit now.

Reviewer: I think you can just remove it, what we want is covered by the Linux Kernel coding style.

Any extra rule will lead to more confusion:
* `txbuf` vs `tx_buf`, both should be ok,
* not sneak camel case macro's,

> * Use `snake case`_ for code and screaming snake case for macros and constants.

Reviewer 3: Do people really know what "screaming" means in this context? Why not use all-caps or similar? Perhaps adding two examples could help?

Author: It was in the wikipedia link, but I clarified it now.

> * Use `snake case`_ for code and screaming snake case for macros and constants.

Reviewer: Please remove this line, a quick `grep -r '#define [a-z]*(' *` on the codebase shows that cannot be a general rule.

Author: What do you mean? This is only for C code, do you mean that we have CamelCase or others in the codebase? I am all for consensus, but Zephyr is snake case, I don't think anyone will dispute that?

Author: Oh I see what you mean now, about the macros. My mistake @Laczen. Let me try and simplify this by just removing the statement about macros.

Author: @Laczen thanks for the comment, I  have now simplified it.

Coding Style
============

.. note::

>    Coding style is enforced on any new or modified code, but contributors are

Reviewer: I think this is wrong for code modifications. When code is modified the resulting final code will not comply with the coding style, so preference should be given to consistency. Reviewers should in this case not only review the changes but also how this new code integrates with the existing code base. In many cases it will be possible to align the new code with the coding style, but not always. A gentle nudge in the direction of improved coding style when possible should be sufficient.

Reviewer 4: "This is not a legally binding contract" wait what? wonder what I signed then :D (joke)

I'm fine with the wording as is as well

Reviewer: If it is intended to be just a guideline, what about reformulating it as:

Coding style is enforced on new additions (drivers, subsystems, ...). For modifications to the existing code base contributors are encouraged to align modifications with the coding style.

Reviewer 2: Disagree, we've been over in a previous comment in this thread, things should not be being added that are wrong just because there is previous code that is wrong, you add more bad code, more technical debt and burden someone else with having to fix it

   not expected to correct the style on existing code that they are not
   modifying.

.. note::
   For style aspects where the guidelines don't offer explicit guidance or

>    permit multiple valid ways to express something, new code must try to follow

Reviewer 5: 2 comments here:
- "new code" doesn't sound like a valid subject in this sentence. Rather than new code trying something, it's probably rather the PR author
- "must" (as in MUST) and "try" doesn't sound to me like they can be used in the same sentence -- this actually feels very much like a synonym for "SHOULD", for if one is expected to "try" to do something, we have to be quite flexible in terms of expectations. What I'm trying to say is that since there is no objective "right" way to implement this recommendation, I don't want it to be something that can justify a "you did it wrong" kind of feedback—especially when the contributor actually feels like they did *try*—let alone an actual nack.

Author: Agree with your objections, I think they are very valid. However, I do think the note serves a purpose, any suggestions as to how to rewrite it?

Reviewer 6: @kartben has a fair point that the verb "try" suggests a human subject, and as such is not the best match for a subject like "the code". I think in this case we could just drop it and have either "must", "should", or "shall" there. 

> What I'm trying to say is that since there is no objective "right" way to implement this recommendation, I don't want it to be something that can justify a "you did it wrong" kind of feedback

I don't think the inability to specify something precisely is a valid reason to dismiss it here. I suppose we could extend this with a requirement for the maintainer/reviewer to be able to explain what the existing pattern/style in the existing code is when requesting such changes. Especially first time contributors to a piece of code are unlikely to see all such conventions/patterns/styles up front, and it's perfectly natural for there to be at least one feedback cycle in reviews to address them.

Author: @kartben how about the proposal from Discord:
>For style aspects where the guidelines don't offer explicit guidance or permit multiple valid ways to express something, contributors should follow the style of existing code in the tree, with higher importance given to "nearby" code (first look at the function, then the same file, then subsystem, etc).

Author: @kartben used the version suggested for now.

>    permit multiple valid ways to express something, new code should try to follow

Reviewer 7: "should" or "shall"?

Reviewer 8: > "should" or "shall"?

Does anyone have an "American English" dictionary? :-)

Reviewer 9: It's not a dialect thing, but in this context since it's followed by "try to", then "should" I think fits, since shall would express determination and contradict the rest of the phrase. Maybe the suggestion here is to replace "should try to" entirely with "shall", I'm not sure I'm what case a style can't be followed and saying to just try is ambiguous as far as how to enforce this guideline

Author: I am fine with "must".

   the style of existing code in the tree, with higher importance given to
   "nearby" code (first look at the function, then the same file, then
   subsystem, etc).

.. _Linux kernel coding style:
   https://kernel.org/doc/html/latest/process/coding-style.html

.. _snake case:
   https://en.wikipedia.org/wiki/Snake_case

In general, follow the `Linux kernel coding style`_, with the following
exceptions and clarifications:

> * Use `snake case`_ for code and variables.

Reviewer 10: but this is covered by linux already, in details, I do not think we need to repeat this here:

https://kernel.org/doc/html/latest/process/coding-style.html#naming

Author: It's not explicit some said, so it was made explicit here.

Reviewer 11: I'm in favor of incremental additions to our style guide that eventually make reference the Linux style guide obsolete.  While unlikely, Linux can change their style guidelines at any time.

Reviewer 10: need to just fork the original linux style into our docs, call them zephyr style guidelines and never look back again.

Reviewer 10: > kernel.org/doc/html/latest/process/coding-style.html#naming

the more I look at this, the more I think we should just go our own and stop referencing this from Linux. It looks more of a rant than guidelines in some parts.

Reviewer 5: > > kernel.org/doc/html/latest/process/coding-style.html#naming
> 
> the more I look at this, the more I think we should just go our own and stop referencing this from Linux. It looks more of a rant than guidelines in some parts.

Yes, totally, and I wanted to point this out the other day. It seriously feels *very* dated in the way it's written, to the point that it's not even reflecting the values we try to promote via our Code of Conduct, for example... Like, doing something wrong being refered to as being a "shooting offense"... Really?! So yeah if there's something where we can clearly state "we're not Linux" maybe it can start with coding style :) big +1 from me

Reviewer 3: Those are very good arguments. I've never before read the Linux coding guidelines like that, but you are absolutely right.

* The line length is 100 columns or fewer. In the documentation, longer lines

>   for URL references are an allowed exception.

Reviewer 5: Given there are quite many places where line length is over 100 columns, does the rule "try to align with style in nearby code" apply? I.e. we're now saying we're not expecting contributors to fix non-conforming code/files, so do we want them to introduce potentially oddly/inconsistently wrapped content to existing files?

Author: Up to others to comment on this, but for me this question really falls on the "case by case" and "common sense" criteria of the reviewers. I don't think rewriting the text or complicating it further will help solve all the corner cases like this one.

Reviewer 2: New code should comply with the requirements

Author: In principle I do not think this text needs changing, but ping me if you disagree @kartben.

Coding Style
============

> .. note::

Reviewer 2: Am I lost here? I'm sure it used to be specifically mentioned that the coding style was snake case but had a grep and can't find it

Author: No, it doesn't say that, but it refers to the Linux coding style guide which does not say it either :) I will add a comment about it.

* The line length is 100 columns or fewer. In the documentation, longer lines
  for URL references are an allowed exception.

> * Variables must be defined at the beginning of a function, in "reverse

Author: I suspect this point may be controversial. If it is I can drop it.

Reviewer 12: I'm afraid we do not follow such convention in many places

Reviewer 6: Also, these are two independent requirements, so even if it was kept it should be split into two. It should also reference scope and not function, since I believe we encourage variables to be defined in the smallest possible scope. Anyway, I also think that this should likely just be dropped. If there's a place for "recommendations" (as opposed to requirements) in this document, maybe the variable ordering (the reverse xmas tree thing) could be mentioned somewhere.

Author: I agree, let's drop this for now, I don't want to delay the current PR with these changes.

Reviewer 13: If possible, would it be possible to set two levels, Required and Advisory, like MISRA-C?
I think that if reverse Christmas tree is made a rule, there will be many cases where it cannot be followed, but I agree with making it clear that it is a desirable norm.
(I know it would be a lot of work...)

* Avoid using binary literals (constants starting with ``0b``).
* Avoid using non-ASCII symbols in code, unless it significantly improves
  clarity, avoid emojis in any case.
* Use proper capitalization of nouns in code comments (e.g. ``UART`` and not
  ``uart`` or ``CMake`` ``cmake``).

> Beyond C code, the following coding style rules apply to other types of files:

Reviewer 8: cool though we should really have some compliance check validation for these to have any hope of being consistent, but hey guess this is the first step

insert: I'm a sign not a cop meme

> Beyond C code, the following coding style rules apply to other types of files:

Reviewer 5: we need some kind of paragraph somewhere very prominently stating that this only applies to new file? or does it? If not, this doesn't suffice to solve the problem of "but I used the exact same style as file X that was the closest thing I found to what I wanted to contribute - how can this be wrong?"

Need something explaining *very* clearly how contributors are expected to deal with files that do *not* comply to these rules, cause in practice this is where a lot of personal preferences tend to pollute the review process ("It must be X because that's what we need it to be going forward!" ... "Of course it needs to be Y unlike what our guidelines say because we want to be consistent with the rest of the file/area/<whatever>", ... etc.)  

Also, many of these rules I hope we can just fix tree-wide if they are that big of an issue for the project (ex. extra blank lines in DT, etc.) otherwise there will ALWAYS be people just copy-pasting from a "broken" file and we will still need to spend cycles talking through why they shouldn't have (and no, "you didn't RTFM and all the fine print in our contribution guidelines!", won't be a good answer IMO)

Author: I am quite sure that we comply with these rules for most CMake, Devicetree and Kconfig files in general. What I can do, as you suggest, is to add a paragraph that states that these rules are for new files (or new code inside a file) only, so that when the user modifies an existing file they don't need to go and fix existing issues that they have not added themselves.

Reviewer 5: > I am quite sure that we comply with these rules for most CMake, Devicetree and Kconfig files in general.

Yeah, well, not really, not even close for some of these rules, hence why I'm bringing it up :) thousands of tabs in CMakeLists.txt files, 100s of extra blank lines in DTS files, ... etc

Author: Regardless of the actual count, the problem does exist. I will update the text to reflect this.

* Avoid using binary literals (constants starting with ``0b``).
* Avoid using non-ASCII symbols in code, unless it significantly improves
  clarity, avoid emojis in any case.

> * Use proper capitalization of nouns in code comments (e.g. ``UART`` and not
>   ``uart`` or ``CMake`` ``cmake``).

Reviewer 3: ```suggestion
* Use proper capitalization of nouns in code comments (e.g. ``UART`` and not
  ``uart``, ``CMake`` and not ``cmake``).
```

Beyond C code, the following coding style rules apply to other types of files:

* CMake

  * Indent with spaces, indentation is two spaces.
  * No space between commands (e.g. ``if``) and the corresponding opening
    bracket (e.g. ``(``).

* Devicetree

  * Indent with tabs.

>   * No additional empty lines after ``};``.

Reviewer 3: Does this mean we cannot have empty lines between node declarations? E.g. like this:
```

node1: node@1 {
     foo;
};
/* No empty line allowed here? */
node2: node@2 {
     bar;
};
```

If so, I don't think that's a good idea.

Author: @nordicjm @gmarull @bjarki-andreasen

Reviewer 2: Hmm it's the wrong wording, it is meant to prevent this:
```
node1: node@1 {
    node2: node@2 {
         bar;
    }; <-- newline gap here is fine

    node2: node@2 {
         bar;
    };
<-- newline gap should not be here
};
```

Reviewer 14: I agree with @henrikbrixandersen, it looks ugly.

Reviewer 4: Maybe change to:
```
* No additional empty lines between a dedented `};``.
```
since
```
/ {
        node {

        };

        node {

        }; <- same level
}; <- dedents
```
dedents

>   * Single empty line between elements.

Reviewer 3: Elements? Nodes? Does this address my concern above? If so, I think we need clearer intention here.

Author: @nordicjm @gmarull @bjarki-andreasen

Reviewer 5: cc devicetree collaborators... @decsny @galak @rruuaanng

Reviewer 4: similarly here, reword to:
```
* Single empty line between same level `};`
```

* Avoid using binary literals (constants starting with ``0b``).
* Avoid using non-ASCII symbols in code, unless it significantly improves
  clarity, avoid emojis in any case.
* Use proper capitalization of nouns in code comments (e.g. ``UART`` and not
  ``uart``, ``CMake`` and not ``cmake``).

Beyond C code, the following coding style rules apply to other types of files:

* CMake

  * Indent with spaces, indentation is two spaces.
  * No space between commands (e.g. ``if``) and the corresponding opening
    bracket (e.g. ``(``).

> * Devicetree

Author: This is actually in the Devicetree spec, unless I am mistaken right @decsny @gmarull @bjarki-andreasen? If so then I will just write "respect the Devicetree specification naming rules".

> * Devicetree

Reviewer 12: Regulators tend to use uppercase, e.g. `LDO1` (following Kernel). I think the list above is actually clear, maybe we should explicitly not recommend underscores for node-name?

  * Indent with tabs.
  * No empty lines before a dedenting ``};``.

>   * Single empty line separating nodes at the same hierarchy level.

Reviewer 2: needs something like `Must have 1 space between item and value` e.g. `dma-channels = <2>;` not `dma-channels=<2>;`

> * Kconfig

Reviewer 11: There's a few additional style items listed at
https://builds.zephyrproject.io/zephyr/pr/83117/docs/build/kconfig/tips.html#header-comments-and-other-nits

We should have one location for the Kconfig style and just have the other locations link to the source of truth.

Author: I will remove that from the Kconfig guide and instead move it here, where the coding style is actually defined.

Reviewer 7: If you do remove them from the Kconfig I suggest you at least add a link there to reference back here.

* Avoid using binary literals (constants starting with ``0b``).
* Avoid using non-ASCII symbols in code, unless it significantly improves
  clarity, avoid emojis in any case.
* Use proper capitalization of nouns in code comments (e.g. ``UART`` and not
  ``uart``, ``CMake`` and not ``cmake``).

Beyond C code, the following coding style rules apply to other types of files:

> * CMake

Reviewer 11: > Thanks for the draft PR and for the proposal. If we go this way though, we should probably then open two more PRs, one for Devicetree and one for Kconfig. And we should also decide whether we want the coding style to be in a single place (as it is in this PR) or in each respective section (as you propose with your draft CMake PR).

I like having one location for coding style - a single location provides is easier to find for new contributors (and for seasoned contributors to reference in PR comments).

> * CMake

Reviewer 10: we can always cross reference them and still list them in one place. i.e. we reference linux styles and do not have them directly in the same page, we can do this with others. As each style guide for each components grows, it will all be one page and become unreadable/maintainable. 
Having saud that, I do not case where they end up, as long as they are complete and more than just the result of cherrypicking.

  * Indent with spaces, indentation is two spaces.

>   * No space between commands (e.g. ``if``) and the corresponding opening

Reviewer 5: ```suggestion
  * Don't use space between commands (e.g. ``if``) and the corresponding opening
```

    bracket (e.g. ``(``).

* Devicetree

  * Indent with tabs.
  * Follow the Devicetree specification conventions and rules.
  * Use dashes (``-``) as word separators for node and property names.

>   * Use underscores (``_``) as word separators in node labels.

Reviewer 11: The term is `nodelabel` not `node label`.

Author: @keith-zephyr I could not find the term `nodelabel` in the spec, nor in our doc except for an example. Am I missing something?

Reviewer 8: it used to be a fairly big source of confusion when label properties were common, probably not so much these days though I see more of those coming back for some reason I don't quite understand

Author: Regardless, those are actually named labels in the spec, and in fact they apply both to nodes and properties:

> The source format allows labels to be attached to any node or property value in the devicetree

Reviewer 8: right so it sounds like they are interchangeable, which they are not

Reviewer 11: I see our documentation refers to the term as *node label*.

For new users, *label* and *node label* are quite easy to confuse.  For consistency with all the DT macros, I do think it's worthwhile to explicitly define *nodelabel` (no space).  But that's outside the scope of this PR.

Author: Made a note, will send a separate PR.

Author: @keith-zephyr https://github.com/zephyrproject-rtos/zephyr/pull/87158

  * Leave a single space on each side of the equal sign (``=``) in property
    definitions.

>   * No empty lines before a dedenting ``};``.

Reviewer 5: ```suggestion
  * Don't insert empty lines before a dedenting ``};``.
```

>   * Single empty line separating nodes at the same hierarchy level.

Reviewer 5: ```suggestion
  * Insert a single empty line to separate nodes at the same hierarchy level.
```

> 

Reviewer 2: seemingly we need something to prevent random newline spacing in the same parts i.e. https://github.com/zephyrproject-rtos/zephyr/pull/83310/files#diff-a188c48a9493b631f516e6672a23b4792710f6c3e7375e7b4956957a9c22d3edR18

Author: I will leave this for a a second pass at it, don't want to extend the scope right now.

Reviewer 15: This is a clear example of being too pick. Having a new line on that part makes easy to ready IMHO.
Why do we need this kind of rules? What we gain with that?

Reviewer 2: > This is a clear example of being too pick. Having a new line on that part makes easy to ready IMHO. Why do we need this kind of rules? What we gain with that?

The files in that PR are a complete mess, compare that with other board files, it's completely at odds with that and doesn't even use the same layout in the same file, changing multiple times. It does not look easy to read

Reviewer 2: From your pr:

> Because makes it more readable.

So to answer:

> Why do we need this kind of rules? What we gain with that?

We need it so people i.e. like you in your PR don't come up with your own styles completely at odds with what is standard in zephyr, we don't need every board or every subsystem to have their own coding styles which add nothing

Author: >Why do we need this kind of rules? What we gain with that?

Because we have had serious disagreements lately between contributors and maintainers. Having as much as possible documented saves us from having more debates around all these topics.

Reviewer 15: First, I don't like the way you start to talk here. I think you are better than this.
Second, you can achieve what you are proposing without been too strict. Have the possibility to group things don't make code worst to read.
Third, nobody thinks the same way and you should respect other opinions, even more when you don't like it. 

I can be wrong, ignore if is that the case, but my impression is that you want to have some other tool than checkpatch. The problem is that if you start to restrict code freedom to achieve your goal that only shows that the tool that you are picking is not capable to do the job. This means that you are doing the tool job instead the contrary.

I have the opinion that people could propose to increase a little the bar on the field they are responsible. Otherwise, it is insane. Imagine that you start to use the same bar on areas that you are not a maintainer. Then the maintainer should respect your opinion on that? Forcing people to do what they don't want never brings good things.

This is why I believe Zephyr should have guidelines that are not too strict.

Reviewer 15: There is no issues about having spaces inside a node. Even mainline Linux have it:

https://github.com/torvalds/linux/blob/fc033cf25e612e840e545f8d5ad2edd6ba613ed5/arch/arm/boot/dts/microchip/at91sam9n12ek.dts#L268-L283

https://github.com/torvalds/linux/blob/fc033cf25e612e840e545f8d5ad2edd6ba613ed5/arch/arm/boot/dts/microchip/sama5d2.dtsi#L44-L58

Author: I don't think the rules say anything about extra empty lines inside a node @nandojve. It only says that you cannot leave an empty line after a closing `};` and that you must leave an empty lines between nodes. But here you are showing empty lines **within** a node and between properties, which is not forbidden at all by the rules above, unless I am mistaken.

Reviewer 15: Hi @carlescufi ,

I'm advocating against below idea from @nordicjm, beginning of this thread. Copying here to easy follow:

> seemingly we need something to prevent random newline spacing in the same parts i.e. https://github.com/zephyrproject-rtos/zephyr/pull/83310/files#diff-a188c48a9493b631f516e6672a23b4792710f6c3e7375e7b4956957a9c22d3edR18

He is using my PR as example of something that project should block. I'm not in favor of his idea and you can see that this pattern exists on mainline Linux too.

So, there is no good reason to be so strict. And this is what I've been saying all the time. IMHO this is not a problem and never will be a problem that we should complain on a review.

Reviewer 2: We should have common styles for files and that looks like this https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/96boards/argonkey/96b_argonkey.dts
No random lines between aliases, no random lines between things inside the same node, **you are making up your own style for your PR which is not how other files are formatted**, something we absolutely should prevent

Reviewer 15: @nordicjm ,

About the new topic that you brought: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/96boards/argonkey/96b_argonkey.dts#L104

You are 100% right about this. I totally agree with you about it. This is not a good way to define a node in devicetree and should be fixed.

> you are making up your own style for your PR which is not how other files are formatted, 

> And to flip the question around, why is it so important/required that your dts files differ from the norm. in zephyr which has been fine for years?

With all respect, I can say, in fact, this is you. We have been using this for many years on Zephyr. The way you are writing you are distorting the facts. This is not different of the norm. I gave an example that this is not only used here but outside, inclusive, and that is a project of some relevance for devicetree.

I think you are simple ignoring the facts to try push your vision. If it is not new, how can I inventing or even pushing something different from the norm?

>  something we absolutely should prevent.

With this you are creating your own style, not me. You want to block something that has been used here and outside for a long time. Why?

I think you are the guy that is asking too much here. I'm saying that we should not touch this part because it is a common practice for many years here and on Linux and don't affect visually the node. The second case that you brought is different scenario and I agreed with you in that case.

Author: Can either of you provide some stats perhaps? if most of the current code complies with one approach or other perhaps we can include it in the list. Otherwise I would rather not tread there to avoid overloading this PR.

Reviewer 3: As I understand it, @nandojve does not want us to introduce guidelines that discourage empty lines between properties within a node. 

I agree. Grouping properties using newlines is common practice in DTS, both in Zephyr and elsewhere. 

I don't currently see anything in the content of this PR for introducing such guidelines, so perhaps we can leave it out of this PR - and address it separately, if anybody desires?

Reviewer 2: > Can either of you provide some stats perhaps? if most of the current code complies with one approach or other perhaps we can include it in the list. Otherwise I would rather not tread there to avoid overloading this PR.

Out of 886 board dts files in zephyr, the following have spaces:
```
boards/mediatek/mt8196/mt8196_adsp.dts line 107
boards/microchip/mec1501modular_assy6885/mec1501modular_assy6885.dts line 25
boards/mxchip/az3166_iotdevkit/az3166_iotdevkit.dts line 19
boards/mxchip/az3166_iotdevkit/az3166_iotdevkit.dts line 23
boards/others/serpente/serpente.dts line 30
```
4 out of 886 = 0.4%

Reviewer 15: Can you put your script here? It seems to be too selective.

https://github.com/zephyrproject-rtos/zephyr/blob/dabeb9c95c719eb325b911c0e325d52b4a9dc244/boards/atmel/sam/sam4l_ek/sam4l_ek.dts#L52-L58

https://github.com/zephyrproject-rtos/zephyr/blob/dabeb9c95c719eb325b911c0e325d52b4a9dc244/boards/atmel/sam/sam_e70_xplained/sam_e70_xplained-common.dtsi#L57-L62

Is this change only valid for the boards directory? Or we are looking only there because @nordicjm is a collaborator of that part.

BTW who is 57300 ?
https://github.com/zephyrproject-rtos/zephyr/blob/dabeb9c95c719eb325b911c0e325d52b4a9dc244/MAINTAINERS.yml#L594-L602

Reviewer 2: As said, only for aliases, https://pastebin.com/tuv8VLsK

> Is this change only valid for the boards directory? Or we are looking only there because @nordicjm is a collaborator of that part.

It's there because that's where board aliases are

* Kconfig

>   * Line length of 100 columns or fewer.

Reviewer 5: ```suggestion
```

Author: Why? This makes sense in the context of Kconfig, because in Devicetree it is not enforced unless I am mistaken?

Reviewer 5: Because I was under the impression that the enclosing section, "Coding style", already applies to everything? Why would this need to be repeated for Kconfig? And why is it not repeated for CMake, then?

Reviewer 5: Reading this again it looks like "Coding Style" is probably meant to read "C coding style", and that you probably didn't mean to fold Kconfig/devicetree/CMake "under" this section as this has really nothing to do with C, and can mistakenly lead people like myself to then interpret "Coding guidelines" as being "General style guidelines" (and e.g. interpret the bullet "In the documentation, longer lines for URL references are an allowed exception." as applying to RST files, whereas I guess the intent was to apply to *Doxygen* documentation?)

Related(ish): The "Coding guidelines" section is pretty buried in the document.

Author: What you say is correct and would benefit from a cleanup, but I am trying really really hard to limit the scope of this PR. If I start now rewriting the Coding Style section I will open a can of worms that I would rather leave closed in this iteration, we can then do another pass cleaning up.

  * Indent with tabs, except for ``help`` entry text which should be placed at
    one tab plus two extra spaces.
  * Leave a single empty line between option declarations.

>   * Statements like ``select`` must be used carefully, see

Reviewer 5: ```suggestion
  * Use statements like ``select`` carefully, see
```
